spring/spring4/spring4-embedded/src/test/java/org/infinispan/spring/provider/SpringCacheTest.java

```diff
@@ -212,6 +210,44 @@ public void testValueLoaderWithNoPreviousValue() {
       assertEquals("test", valueFromCache.get());
    }
 
+   /*
+    * In this test Thread 1 should exclusively block Cache#get method so that Thread 2 won't be able to
+    * insert "thread2" string into the cache.
+    *
+    * The test check this part of the Spring spec:
+    * Return the value to which this cache maps the specified key, obtaining that value from valueLoader if necessary.
+    * This method provides a simple substitute for the conventional "if cached, return; otherwise create, cache and return" pattern.
+    * @see http://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/cache/Cache.html#get-java.lang.Object-java.util.concurrent.Callable-
+    */
+   @Test(timeOut = 30_000, invocationCount = 100)
+   public void testValueLoaderWithLocking() throws Exception {
+      //given
+      CountDownLatch waitUntilThread1LocksValueGetter = new CountDownLatch(1);
+
+      //when
+      Future<String> thread1 = fork(() -> cache.get("test", () -> {
+         waitUntilThread1LocksValueGetter.countDown();
+         TimeUnit.MILLISECONDS.sleep(10);
+         return "thread1";
+      }));
+
+      Future<String> thread2 = fork(() -> {
+         waitUntilThread1LocksValueGetter.await();
```

Comment:
@rvansa indeed. `AbstractInfinispanTest.killSpawnedThreads()` will eventually kill the thread, but it's better to make tests as explicit as possible.

In fact I avoid `@Test(timeout = ...)` as much as I can, just like `@Test(expectedExceptions = ...)`, because they're too coarse and it's hard for the reader to figure out which of the statements you are expecting to block/throw an exception. The fact that TestNG hasn't yet figured out how to run such test methods in the same thread as their configuration methods (requiring `TestResourceTracker.testThreadStarted()` calls before starting a cache manager) makes me want to use them even less.